modules/apps/portal/portal-upgrade-test/src/testIntegration/java/com/liferay/portal/upgrade/v7_4_x/test/UpgradeDLFileShortcutExternalReferenceCodeTest.java

```diff
@@ -0,0 +1,92 @@
+ */
+
+package com.liferay.portal.upgrade.v7_4_x.test;
+
+import com.liferay.arquillian.extension.junit.bridge.junit.Arquillian;
+import com.liferay.document.library.kernel.model.DLFileEntry;
+import com.liferay.document.library.kernel.model.DLFileShortcut;
+import com.liferay.document.library.kernel.model.DLFolder;
+import com.liferay.document.library.kernel.model.DLFolderConstants;
+import com.liferay.document.library.kernel.service.DLFileShortcutLocalService;
+import com.liferay.document.library.test.util.DLTestUtil;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.model.ExternalReferenceCodeModel;
+import com.liferay.portal.kernel.test.util.ServiceContextTestUtil;
+import com.liferay.portal.kernel.test.util.TestPropsValues;
+import com.liferay.portal.kernel.version.Version;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.upgrade.registry.UpgradeStepRegistrator;
+import com.liferay.portal.upgrade.test.util.BaseExternalReferenceCodeUpgradeProcessTestCase;
+import com.liferay.portal.upgrade.v7_4_x.UpgradeDLFileShortcutExternalReferenceCode;
+
+import org.junit.runner.RunWith;
+
+/**
+ * @author Balázs Sáfrány-Kovalik
+ */
+@RunWith(Arquillian.class)
+public class UpgradeDLFileShortcutExternalReferenceCodeTest
+	extends BaseExternalReferenceCodeUpgradeProcessTestCase {
+
+	@Override
+	protected ExternalReferenceCodeModel[] addExternalReferenceCodeModels(
+			String tableName)
+		throws PortalException {
+
+		DLFolder dlFolder = DLTestUtil.addDLFolder(group.getGroupId());
+
+		DLFileEntry dlFileEntry = DLTestUtil.addDLFileEntry(
+			dlFolder.getFolderId());
+
+		DLFileShortcut dlFileShortcut =
+			_dlFileShortcutLocalService.addFileShortcut(
+				null, TestPropsValues.getUserId(), TestPropsValues.getGroupId(),
+				TestPropsValues.getGroupId(),
+				DLFolderConstants.DEFAULT_PARENT_FOLDER_ID,
+				dlFileEntry.getFileEntryId(),
+				ServiceContextTestUtil.getServiceContext(
+					TestPropsValues.getGroupId(), TestPropsValues.getUserId()));
+
+		return new ExternalReferenceCodeModel[] {dlFileShortcut};
+	}
+
+	@Override
+	protected ExternalReferenceCodeModel fetchExternalReferenceCodeModel(
+			ExternalReferenceCodeModel externalReferenceCodeModel,
+			String tableName)
+		throws PortalException {
+
+		DLFileShortcut dlFileShortcut =
+			(DLFileShortcut)externalReferenceCodeModel;
+
+		return _dlFileShortcutLocalService.fetchDLFileShortcut(
+			dlFileShortcut.getFileShortcutId());
+	}
+
+	@Override
+	protected String[] getTableNames() {
+		return new String[] {"DLFileShortcut"};
+	}
+
+	@Override
+	protected UpgradeDLFileShortcutExternalReferenceCode getUpgradeProcess() {
+		return new UpgradeDLFileShortcutExternalReferenceCode();
+	}
+
+	@Override
+	protected UpgradeStepRegistrator getUpgradeStepRegistrator() {
+		return null;
+	}
+
+	@Override
+	protected Version getVersion() {
+		return null;
+	}
```

Comment:
Why not implement these methods and reuse the `getUpgradeProcess` logic in the base class? See other ERC upgrade classes.

Reply:
Hey,
I've overridden getUpgradeProcess() to instantiate the class-under-test UpgradeDLFileShortcutExternalReferenceCode in order to avoid a brittle dependency on a Version number identified by "magic Strings". Should the upgrade process' version change in the register, I'd have to update the test as well.

I must have found a null implementation as example such as [LayoutExternalReferenceCodeUpgradeProcessTest](https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal/portal-upgrade-test/src/testIntegration/java/com/liferay/portal/upgrade/v7_4_x/test/LayoutExternalReferenceCodeUpgradeProcessTest.java#L74) or [DLFileEntryTypeExternalReferenceCodeUpgradeProcessTest](https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal/portal-upgrade-test/src/testIntegration/java/com/liferay/portal/upgrade/v7_4_x/test/DLFileEntryTypeExternalReferenceCodeUpgradeProcessTest.java#L108)

Reply:
Ok. The only thing is that all other ERC upgrades reference the version instead of instantiating the upgrade step themselves, so there's some risk that Brian will not like it.

But as it looks correct, I'll let you decide.